### DISCUSSION TITLE
How to subscribe events in nested ChildRegion

### DISCUSSION THREAD

**Discussion by ngoquoctoandev** [PrismLibrary/Prism#2506] [WPF]
![Untitled](https://user-images.githubusercontent.com/78490340/124377471-485b7f00-dcd6-11eb-919c-10973a64dac3.png)

I have a form designed like this:
**MainWindow** contains a **TextBlock** and a **ContentRegion**.
I registered **ViewA** to **ContentRegion** when MainWindow is initialized.
On ViewA I designed two buttons to call ViewB, or ViewC into **ChildRegion** of ViewA.
I publish an event at the **OnStartup** method to send a message to all ViewModels that subscribe to it.
**Scenario 1**: ViewA receives the event while ViewB, ViewC do not receive this event. But it can only call ViewB, ViewC into **ChildRegion** of ViewA. Obviously now in **RegionManager** only contains **ContentRegion**, does not contain **ChildRegion**, so the event registered at ViewB, ViewC will not receive.
Here is the code I wrote:

```
public class ModuleNameModule : IModule
{
    private readonly IRegionManager _regionManager;

    public ModuleNameModule(IRegionManager regionManager) => _regionManager = regionManager;

    public void OnInitialized(IContainerProvider containerProvider)
    {
        _regionManager.RequestNavigate(RegionNames.ContentRegion, nameof(ViewA));
    }

    public void RegisterTypes(IContainerRegistry containerRegistry)
    {
        containerRegistry.RegisterForNavigation<ViewA>();
        containerRegistry.RegisterForNavigation<ViewB>();
        containerRegistry.RegisterForNavigation<ViewC>();
    }
}
```
And this is how I call ViewB, ViewC into ViewA's ChildRegion

![image](https://user-images.githubusercontent.com/78490340/124378133-092f2d00-dcda-11eb-918c-0ed361d6c1e9.png)

Scenario 2: I assign object of IRegion to constructor of ViewAViewModel, but error occurs because RegionManager does not contain ChildRegion.

![image](https://user-images.githubusercontent.com/78490340/124378467-a0e14b00-dcdb-11eb-9b00-d914de99fb25.png)

![image](https://user-images.githubusercontent.com/78490340/124378077-bb1a2980-dcd9-11eb-83ca-f1b1095f9220.png)

Is there any solution to register ChildRegion on initialization to receive events at ViewB, ViewC as soon as program is showed?

This is my source code, can anyone help me to receive events at ViewB, ViewC as soon as the program appears?
https://github.com/z2hteam/PrismWpfDemo.git

**Comment by brianlagunas** [comment]
First off, don't use Shared projects with EventAggregator. Shared projects are compiled into the referenced project and seen as a separate class. Meaning your messages will not work because the Event class will not be seen as the same event class across your various projects.

You also can't send message to objects that don't exist, and you can't add views to a region until it exists. You'll have to make sure you understand the lifecycle of your application to implement this in your app.

**Reply by ngoquoctoandev** [reply]
because i make a dynamic menu. so need to ask the problem like above, is there any way to solve it in the best way?

**Reply by brianlagunas** [reply]
Of course you can solve your problem. What you are trying to do is very common. You just need to spend more time learning the framework so you know how to best implement your solution.

Maybe my course will help: http://bit.ly/PrismForWpf
